packages/backend-common/src/reading/GcsUrlReader.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Storage } from '@google-cloud/storage';
+import {
+  ReaderFactory,
+  ReadTreeResponse,
+  SearchResponse,
+  UrlReader,
+} from './types';
+import getRawBody from 'raw-body';
+import { Logger } from 'winston';
+
+const parseURL = (
+  url: string,
+): { host: string; bucket: string; key: string } => {
+  const { host, pathname } = new URL(url);
+
+  if (host !== 'storage.cloud.google.com') {
+    throw new Error(`not a valid GCS URL: ${url}`);
+  }
+
+  const [, bucket, ...key] = pathname.split('/');
+  return {
+    host: host,
+    bucket,
+    key: key.join('/'),
+  };
+};
+
+export class GcsUrlReader implements UrlReader {
+  static factory: ReaderFactory = ({ config, logger }) => {
+    if (!config.has('integrations.gcs')) {
+      return [];
+    }
+    return config
+      .getConfigArray('integrations.gcs')
+      .filter(integration => {
+        if (!integration.has('clientEmail') || !integration.has('privateKey')) {
+          logger.warn(
+            "Skipping gcs integration, Missing required config value at 'integration.gcs.clientEmail' or 'integration.gcs.privateKey'",
+          );
+          return false;
+        }
+        return true;
+      })
+      .map(integration => {
+        const privKey = integration
+          .getOptionalString('privateKey')
+          ?.split('\\n')
+          .join('\n');
+
+        const storage = new Storage({
+          credentials: {
+            client_email: integration.getOptionalString('clientEmail'),
+            private_key: privKey,
+          },
+        });
```

Comment:
Should be possible to optionally provide credentials if available in `app-config.yaml`, otherwise let the client authenticate using Environment variable.

https://github.com/backstage/backstage/blob/8504934ae964a0a2d7261f5198902a6cf138e0bc/packages/techdocs-common/src/stages/publish/googleStorage.ts#L62-L66

Reply:
I haven't done this in the exact same idiomatic way but now it is possible to use the default credential discovery. Let me know if there is a better way to rewrite that :)